File: CS/DevExpressMvcApplication1/Models/Northwind.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Data;
using System.Data.Linq;
using System.Linq;
using System.Web;
using System.Web.UI;


    public static class NorthwindDataProvider {
        const string NorthwindDataContextKey = "DXNorthwindDataContext";

        public static NorthwindDataContext DB {
            get {
                if(HttpContext.Current.Items[NorthwindDataContextKey] == null)
                    HttpContext.Current.Items[NorthwindDataContextKey] = new NorthwindDataContext();
                return (NorthwindDataContext)HttpContext.Current.Items[NorthwindDataContextKey];
            }
        }

        static double CalculateAveragePrice(int categoryID) {
            return (double)(from product in DB.Products where product.CategoryID == categoryID select product).Average(s => s.UnitPrice);
        }
        public static IEnumerable GetCategories() {
            return from category in DB.Categories select category;
        }
        public static IEnumerable GetCategoriesNames() {
            return from category in DB.Categories select category.CategoryName;
        }
        public static IEnumerable GetCategoriesAverage() {
            return from category in DB.Categories select new {
                category.CategoryName,
                AvgPrice = CalculateAveragePrice(category.CategoryID)
            };
        }
        public static IEnumerable GetCustomers() {
            return from customer in DB.Customers select customer;
        }
        public static IEnumerable GetProducts() {
            return from product in DB.Products select product;
        }
        public static IEnumerable GetProducts(string categoryName) {
            return from product in DB.Products
                   join category in DB.Categories on product.CategoryID equals category.CategoryID
                   where category.CategoryName == categoryName
                   select product;
        }
        public static IEnumerable GetEmployees() {
            return from employee in DB.Employees select employee;
        }
        public static Binary GetEmployeePhoto(int employeeId) {
            return (from employee in DB.Employees
                    where employee.EmployeeID == employeeId
                    select employee.Photo).SingleOrDefault();
        }
        public static string GetEmployeeNotes(int employeeId) {
            return (from employee in DB.Employees
                    where employee.EmployeeID == employeeId
                    select employee.Notes).Single();
        }
        public static IEnumerable GetOrders() {
            return from order in DB.Orders select order;
        }
        public static IEnumerable GetInvoices() {
            return from invoice in DB.Invoices
                   join customer in DB.Customers on invoice.CustomerID equals customer.CustomerID
                   select new {
                       customer.CompanyName,
                       invoice.City,
                       invoice.Region,
                       invoice.Country,
                       invoice.UnitPrice,
                       invoice.Quantity
                   };
        }
        public static IEnumerable GetFullInvoices() {
            return from invoice in DB.Invoices
                   join customer in DB.Customers on invoice.CustomerID equals customer.CustomerID
                   join order in DB.Orders on invoice.OrderID equals order.OrderID
                   select new {
                       SalesPerson = order.Employee.FirstName + " " + order.Employee.LastName,
                       customer.CompanyName,
                       invoice.Country,
                       invoice.Region,
                       invoice.OrderDate,
                       invoice.ProductName,
                       invoice.UnitPrice,
                       invoice.Quantity
                   };
        }
        public static IEnumerable GetInvoices(string customerID) {
            return from invoice in DB.Invoices where invoice.CustomerID == customerID select invoice;
        }
        
        public static IList<EditableProduct> GetEditableProducts() {
            IList<EditableProduct> products = (IList<EditableProduct>)HttpContext.Current.Session["Products"];

            if(products == null) {
                products = (from product in DB.Products
                            select new EditableProduct {
                                ProductID = product.ProductID,
                                ProductName = product.ProductName,
                                CategoryID = product.CategoryID,
                                QuantityPerUnit = product.QuantityPerUnit,
                                UnitPrice = product.UnitPrice,
                                UnitsInStock = product.UnitsInStock,
                                Discontinued = product.Discontinued
                            }).ToList();
                HttpContext.Current.Session["Products"] = products;
            }
            return products;
        }
        public static EditableProduct GetEditableProduct(int productID) {
            return (from product in GetEditableProducts() where product.ProductID == productID select product).FirstOrDefault();
        }
        public static int GetNewEditableProductID() {
            EditableProduct lastProduct = (from product in GetEditableProducts() select product).Last();
            return (lastProduct != null) ? lastProduct.ProductID + 1 : 0;
        }
        public static void DeleteProduct(int productID) {
            EditableProduct product = GetEditableProduct(productID);
            if(product != null)
                GetEditableProducts().Remove(product);
        }
        public static void InsertProduct(EditableProduct product) {
            EditableProduct editProduct = new EditableProduct();
            editProduct.ProductID = GetNewEditableProductID();
            editProduct.ProductName = product.ProductName;
            editProduct.CategoryID = product.CategoryID;
            editProduct.QuantityPerUnit = product.QuantityPerUnit;
            editProduct.UnitPrice = product.UnitPrice;
            editProduct.UnitsInStock = product.UnitsInStock;
            editProduct.Discontinued = product.Discontinued;
            GetEditableProducts().Add(editProduct);
        }
        public static void UpdateProduct(EditableProduct product) {
            EditableProduct editProduct = GetEditableProduct(product.ProductID);
            if(editProduct != null) {
                editProduct.ProductName = product.ProductName;
                editProduct.CategoryID = product.CategoryID;
                editProduct.QuantityPerUnit = product.QuantityPerUnit;
                editProduct.UnitPrice = product.UnitPrice;
                editProduct.UnitsInStock = product.UnitsInStock;
                editProduct.Discontinued = product.Discontinued;
            }
        }
        
        public static IEnumerable GetEmployeesList() {
            return from employee in DB.Employees
                   select new {
                       ID = employee.EmployeeID,
                       Name = employee.LastName + " " + employee.FirstName
                   };
        }
        public static int GetFirstEmployeeID() {
            return (from employee in DB.Employees
                    select employee.EmployeeID).First<int>();
        }
        public static Employee GetEmployee(int employeeId) {
            return (from employee in DB.Employees
                    where employeeId == employee.EmployeeID
                    select employee).Single<Employee>();
        }
        public static IEnumerable GetOrders(int employeeID) {
            return from order in DB.Orders
                   where order.EmployeeID == employeeID
                   join order_detail in DB.Order_Details on order.OrderID equals order_detail.OrderID
                   join customer in DB.Customers on order.CustomerID equals customer.CustomerID
                   select new {
                       order.OrderID,
                       order.ShipName,
                       order_detail.Quantity,
                       order_detail.UnitPrice,
                       customer.ContactName,
                       customer.CompanyName,
                       customer.City,
                       customer.Address,
                       customer.Phone,
                       customer.Fax
                   };
        }
    }

    public class EditableProduct {
        public int ProductID { get; set; }

        [Required(ErrorMessage = "Product Name is required")]
        [StringLength(50, ErrorMessage = "Must be under 50 characters")]
        public string ProductName { get; set; }

        [Required(ErrorMessage = "Category is required")]
        public int? CategoryID { get; set; }

        [StringLength(100, ErrorMessage = "Must be under 100 characters")]
        public string QuantityPerUnit { get; set; }

        [Range(0, 10000, ErrorMessage = "Must be between 0 and 10000$")]
        public decimal? UnitPrice { get; set; }

        [Range(0, 1000, ErrorMessage = "Must be between 0 and 1000")]
        public short? UnitsInStock { get; set; }

        bool? discontinued;
        public bool? Discontinued {
            get {
                return discontinued;
            }
            set {
                discontinued = value == null ? false : value;
            }
        }
    }

File: VB/DevExpressMvcApplication1/Models/Northwind.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.ComponentModel.DataAnnotations
Imports System.Data
Imports System.Data.Linq
Imports System.Linq
Imports System.Web
Imports System.Web.UI


	Public NotInheritable Class NorthwindDataProvider
		Private Const NorthwindDataContextKey As String = "DXNorthwindDataContext"

		Private Sub New()
		End Sub
		Public Shared ReadOnly Property DB() As NorthwindDataContext
			Get
				If HttpContext.Current.Items(NorthwindDataContextKey) Is Nothing Then
					HttpContext.Current.Items(NorthwindDataContextKey) = New NorthwindDataContext()
				End If
				Return CType(HttpContext.Current.Items(NorthwindDataContextKey), NorthwindDataContext)
			End Get
		End Property

		Private Shared Function CalculateAveragePrice(ByVal categoryID As Integer) As Double
			Return CDbl(( _
						From product In DB.Products _
						Where product.CategoryID.Equals(categoryID) _
						Select product).Average(Function(s) s.UnitPrice))
		End Function
		Public Shared Function GetCategories() As IEnumerable
			Return _
				From category In DB.Categories _
				Select category
		End Function
		Public Shared Function GetCategoriesNames() As IEnumerable
			Return _
				From category In DB.Categories _
				Select category.CategoryName
		End Function
		Public Shared Function GetCategoriesAverage() As IEnumerable
			Return _
				From category In DB.Categories _
				Select New With {Key category.CategoryName, Key .AvgPrice = CalculateAveragePrice(category.CategoryID)}
		End Function
		Public Shared Function GetCustomers() As IEnumerable
			Return _
				From customer In DB.Customers _
				Select customer
		End Function
		Public Shared Function GetProducts() As IEnumerable
			Return _
				From product In DB.Products _
				Select product
		End Function
		Public Shared Function GetProducts(ByVal categoryName As String) As IEnumerable
			Return _
				From product In DB.Products _
				Join category In DB.Categories On product.CategoryID Equals category.CategoryID _
				Where category.CategoryName = categoryName _
				Select product
		End Function
		Public Shared Function GetEmployees() As IEnumerable
			Return _
				From employee In DB.Employees _
				Select employee
		End Function
		Public Shared Function GetEmployeePhoto(ByVal employeeId As Integer) As Binary
			Return ( _
					From employee In DB.Employees _
					Where employee.EmployeeID.Equals(employeeId) _
					Select employee.Photo).SingleOrDefault()
		End Function
		Public Shared Function GetEmployeeNotes(ByVal employeeId As Integer) As String
			Return ( _
					From employee In DB.Employees _
					Where employee.EmployeeID.Equals(employeeId) _
					Select employee.Notes).Single()
		End Function
		Public Shared Function GetOrders() As IEnumerable
			Return _
				From order In DB.Orders _
				Select order
		End Function
		Public Shared Function GetInvoices() As IEnumerable
			Return _
				From invoice In DB.Invoices _
				Join customer In DB.Customers On invoice.CustomerID Equals customer.CustomerID _
				Select New With {Key customer.CompanyName, Key invoice.City, Key invoice.Region, Key invoice.Country, Key invoice.UnitPrice, Key invoice.Quantity}
		End Function
		Public Shared Function GetFullInvoices() As IEnumerable
			Return _
				From invoice In DB.Invoices _
				Join customer In DB.Customers On invoice.CustomerID Equals customer.CustomerID _
				Join order In DB.Orders On invoice.OrderID Equals order.OrderID _
				Select New With {Key .SalesPerson = order.Employee.FirstName & " " & order.Employee.LastName, Key customer.CompanyName, Key invoice.Country, Key invoice.Region, Key invoice.OrderDate, Key invoice.ProductName, Key invoice.UnitPrice, Key invoice.Quantity}
		End Function
		Public Shared Function GetInvoices(ByVal customerID As String) As IEnumerable
			Return _
				From invoice In DB.Invoices _
				Where invoice.CustomerID = customerID _
				Select invoice
		End Function

		Public Shared Function GetEditableProducts() As IList(Of EditableProduct)
			Dim products As IList(Of EditableProduct) = CType(HttpContext.Current.Session("Products"), IList(Of EditableProduct))

			If products Is Nothing Then
				products = ( _
						From product In DB.Products _
						Select New EditableProduct With {.ProductID = product.ProductID, .ProductName = product.ProductName, .CategoryID = product.CategoryID, .QuantityPerUnit = product.QuantityPerUnit, .UnitPrice = product.UnitPrice, .UnitsInStock = product.UnitsInStock, .Discontinued = product.Discontinued}).ToList()
				HttpContext.Current.Session("Products") = products
			End If
			Return products
		End Function
		Public Shared Function GetEditableProduct(ByVal productID As Integer) As EditableProduct
			Return ( _
					From product In GetEditableProducts() _
					Where product.ProductID = productID _
					Select product).FirstOrDefault()
		End Function
		Public Shared Function GetNewEditableProductID() As Integer
			Dim lastProduct As EditableProduct = ( _
					From product In GetEditableProducts() _
					Select product).Last()
			Return If((lastProduct IsNot Nothing), lastProduct.ProductID + 1, 0)
		End Function
		Public Shared Sub DeleteProduct(ByVal productID As Integer)
			Dim product As EditableProduct = GetEditableProduct(productID)
			If product IsNot Nothing Then
				GetEditableProducts().Remove(product)
			End If
		End Sub
		Public Shared Sub InsertProduct(ByVal product As EditableProduct)
			Dim editProduct As New EditableProduct()
			editProduct.ProductID = GetNewEditableProductID()
			editProduct.ProductName = product.ProductName
			editProduct.CategoryID = product.CategoryID
			editProduct.QuantityPerUnit = product.QuantityPerUnit
			editProduct.UnitPrice = product.UnitPrice
			editProduct.UnitsInStock = product.UnitsInStock
			editProduct.Discontinued = product.Discontinued
			GetEditableProducts().Add(editProduct)
		End Sub
		Public Shared Sub UpdateProduct(ByVal product As EditableProduct)
			Dim editProduct As EditableProduct = GetEditableProduct(product.ProductID)
			If editProduct IsNot Nothing Then
				editProduct.ProductName = product.ProductName
				editProduct.CategoryID = product.CategoryID
				editProduct.QuantityPerUnit = product.QuantityPerUnit
				editProduct.UnitPrice = product.UnitPrice
				editProduct.UnitsInStock = product.UnitsInStock
				editProduct.Discontinued = product.Discontinued
			End If
		End Sub

		Public Shared Function GetEmployeesList() As IEnumerable
			Return _
				From employee In DB.Employees _
				Select New With {Key .ID = employee.EmployeeID, Key .Name = employee.LastName & " " & employee.FirstName}
		End Function
		Public Shared Function GetFirstEmployeeID() As Integer
			Return ( _
					From employee In DB.Employees _
					Select employee.EmployeeID).First()
		End Function
		Public Shared Function GetEmployee(ByVal employeeId As Integer) As Employee
			Return ( _
					From employee In DB.Employees _
					Where employeeId.Equals(employee.EmployeeID) _
					Select employee).Single()
		End Function
		Public Shared Function GetOrders(ByVal employeeID As Integer) As IEnumerable
			Return _
				From order In DB.Orders _
				Where order.EmployeeID.Equals(employeeID) _
				Join order_detail In DB.Order_Details On order.OrderID Equals order_detail.OrderID _
				Join customer In DB.Customers On order.CustomerID Equals customer.CustomerID _
				Select New With {Key order.OrderID, Key order.ShipName, Key order_detail.Quantity, Key order_detail.UnitPrice, Key customer.ContactName, Key customer.CompanyName, Key customer.City, Key customer.Address, Key customer.Phone, Key customer.Fax}
		End Function
	End Class

	Public Class EditableProduct
		Private privateProductID As Integer
		Public Property ProductID() As Integer
			Get
				Return privateProductID
			End Get
			Set(ByVal value As Integer)
				privateProductID = value
			End Set
		End Property

		Private privateProductName As String
		<Required(ErrorMessage := "Product Name is required"), StringLength(50, ErrorMessage := "Must be under 50 characters")> _
		Public Property ProductName() As String
			Get
				Return privateProductName
			End Get
			Set(ByVal value As String)
				privateProductName = value
			End Set
		End Property

		Private privateCategoryID? As Integer
		<Required(ErrorMessage := "Category is required")> _
		Public Property CategoryID() As Integer?
			Get
				Return privateCategoryID
			End Get
			Set(ByVal value? As Integer)
				privateCategoryID = value
			End Set
		End Property

		Private privateQuantityPerUnit As String
		<StringLength(100, ErrorMessage := "Must be under 100 characters")> _
		Public Property QuantityPerUnit() As String
			Get
				Return privateQuantityPerUnit
			End Get
			Set(ByVal value As String)
				privateQuantityPerUnit = value
			End Set
		End Property

		Private privateUnitPrice? As Decimal
		<Range(0, 10000, ErrorMessage := "Must be between 0 and 10000$")> _
		Public Property UnitPrice() As Decimal?
			Get
				Return privateUnitPrice
			End Get
			Set(ByVal value? As Decimal)
				privateUnitPrice = value
			End Set
		End Property

		Private privateUnitsInStock? As Short
		<Range(0, 1000, ErrorMessage := "Must be between 0 and 1000")> _
		Public Property UnitsInStock() As Short?
			Get
				Return privateUnitsInStock
			End Get
			Set(ByVal value? As Short)
				privateUnitsInStock = value
			End Set
		End Property

		Private discontinued_Renamed? As Boolean
		Public Property Discontinued() As Boolean?
			Get
				Return discontinued_Renamed
			End Get
			Set(ByVal value? As Boolean)
				discontinued_Renamed = If(value Is Nothing, False, value)
			End Set
		End Property
	End Class

